### PR TITLE
fix(crypto): harden shielded burn ovk encryption

### PR DESCRIPTION
**Pull request by Federico2014**
**What does this PR do?**

Hardens the `out_ciphertext` (ovk-encrypted note) used by shielded TRC-20 `burn`:

- `NoteEncryption.encryptBurnMessageByOvk` generates a fresh CSPRNG nonce per call and emits a 96-byte record laid out as `cipher(80) || nonce(12) || reserved(4)`. The 2^-96 all-zero draw is rerolled so the nonce field never collides with the pre-upgrade zero padding at the same offset.
- `decryptBurnMessageByOvk(ovk, ciphertext, nonce)` requires a 12-byte nonce; null or wrong-length input is rejected via a fail-fast precondition. Pre-upgrade burn logs carry an all-zero byte field at the nonce offset and decrypt via the same path.
- `ShieldedTRC20ParametersBuilder` emits the 96-byte `burnCiphertext`.
- `Wallet.getNoteTxFromLogListByOvk` parses the nonce slice from on-chain logs and feeds it into the new decrypt API.
- `Wallet.getTriggerInputForShieldedTRC20Contract` accepts only the 96-byte record. Stale clients passing an 80-byte cipher are rejected with a clear error message guiding them to rebuild against the new encoder; padding stale ciphertext to a zero-nonce record would otherwise extend the (ovk, zero-nonce) reuse surface forward in time.

**Why are these changes required?**

The previous burn AEAD reused a fixed all-zero nonce under the same `ovk` for every burn, violating RFC 8439 §4 (key, nonce) uniqueness. Two burns under the same `ovk` recover the keystream and the Poly1305 one-time key, which is sufficient to forge AEAD ciphertexts attributed to that ovk.

**Compatibility matrix**

| Scenario | Behavior |
|---|---|
| New `create` + new `scan` | Works. Each burn uses a fresh CSPRNG nonce; (ovk, nonce) reuse is gone. |
| Pre-upgrade burn log + new `scan` | Works. The nonce slice is all-zero on legacy records and decrypts via the same path. |
| New `create` + old `scan` | **Incompatible.** Old scanners always pass a zero nonce to ChaCha20-Poly1305, so they cannot decrypt new burns whose nonce is random. |
| Pre-upgrade 80-byte `burnCiphertext` fed into the new `getTriggerInputForShieldedTRC20Contract` | Rejected (`Wallet.java:4291`) with an explicit error directing the caller to rebuild against `encryptBurnMessageByOvk`. Padding stale ciphertext to a zero-nonce record would extend the (ovk, zero-nonce) reuse window forward in time. |
| New 96-byte record fed into an old `getTriggerInputForShieldedTRC20Contract` | Rejected by old nodes (their length check accepts only 80 bytes). |
| Old node / old client keeps producing zero-nonce burns on-chain | New scanners still decrypt them, since the scan path must remain compatible with historical records. The (ovk, zero-nonce) reuse surface persists for legacy producers until they upgrade — see Stage 2 follow-up below. |

**This PR has been tested by:**
- Unit Tests
  - `NoteEncDecryTest`: round-trip with random nonce, uniqueness across calls, pre-upgrade zero-nonce decode, malformed nonce rejection, short-log rejection, 80-byte trigger-input rejection, 96-byte trigger-input round-trip.
- Manual Testing
  - Local node: regenerate burn parameters end-to-end and verify scan APIs decode pre-upgrade and new records.

**Follow up**

A scanner-side activation height (`BURN_NONCE_ACTIVATION_HEIGHT`, default `Long.MAX_VALUE`, switchable via a `ChainParameters` proposal) is planned as a separate change. Once activated, `getNoteTxFromLogListByOvk` will refuse to decrypt any zero-nonce burn whose log height is ≥ H₀, closing the residual phantom-burn injection path (Finding 1) that bypasses the encoder API.

Note: the gate is a hard reject and cannot distinguish injected phantom burns from legitimate burns produced by un-upgraded clients. Activating it therefore requires an ecosystem migration window: every producer that bypasses `getTriggerInputForShieldedTRC20Contract` (legacy SDKs / wallets / third-party scripts) must be on the 96-byte encoder before H₀, otherwise their post-H₀ burns become unscannable. Tracking in the design doc.

**Extra details**

- No consensus / hard-fork impact: change is confined to off-chain encoding and parsing; zk-SNARK circuit and on-chain validation are unchanged.
- No DB / config changes.

## Summary by CodeRabbit

* **Bug Fixes**
  * Stricter validation for shielded TRC-20 burn payloads: rejects legacy 80-byte burn ciphertexts and enforces exact record size, reducing ambiguous/invalid decrypts.
  * Nonce-aware burn-message decryption now validates decrypted amount and recipient before exposing transparent fields.

* **New Features**
  * Burn records now include a per-burn nonce for improved encryption semantics.

* **Tests**
  * Added tests for round-trip encryption/decryption, nonce uniqueness, legacy rejection, and edge-case log handling.

### CODE REVIEW

File: framework/src/main/java/org/tron/core/zen/note/NoteEncryption.java [hunk 1]
```diff
@@ -265,28 +281,46 @@ public static Optional<byte[]> encryptBurnMessageByOvk(byte[] ovk, BigInteger to
...
+     * recovered from the log; if null, empty, or all-zero (legacy v1 records written before
+     * the random-nonce upgrade), the legacy zero-nonce path is used so existing burns remain
+     * decryptable.
      */
-    public static Optional<byte[]> decryptBurnMessageByOvk(byte[] ovk, byte[] ciphertext)
-        throws ZksnarkException {
+    public static Optional<byte[]> decryptBurnMessageByOvk(byte[] ovk, byte[] ciphertext,
+        byte[] nonce) throws ZksnarkException {
+      byte[] effectiveNonce = (nonce != null && nonce.length == BURN_NONCE_LEN)
+          ? nonce : new byte[BURN_NONCE_LEN];
```
[the remark this answers is not in the document]
Comment (Federico2014) [reply framework/src/main/java/org/tron/core/zen/note/NoteEncryption.java]: The only production caller (`Wallet.getNoteTxFromLogListByOvk`) does a length check on the log and then extracts the nonce as a fixed-length 12B slice, so a `nonce.length != BURN_NONCE_LEN` value cannot reach this method in practice. Rather than guard a path that can't be hit, the follow-up commit removes the fallback entirely — the method now uses the supplied `nonce` directly, and the legacy v1 records still decrypt because their pre-upgrade zero padding *is* a valid 12B all-zero nonce. Thanks for the flag.

[the remark this answers is not in the document]
Comment (Federico2014) [reply framework/src/main/java/org/tron/core/zen/note/NoteEncryption.java]: Good call — added the precondition guard at the top of the 3-arg overload (`if (nonce == null || nonce.length != BURN_NONCE_LEN) return Optional.empty();`) plus a unit test covering wrong-length and null nonce. Will go in with the follow-up commit.
